test/integration/job/job_test.go

```diff
@@ -47,19 +47,167 @@ import (
+		completionMode                  batchv1.CompletionMode
+		wantJobFinishedNumMetricValues  []metricLabelMapWithValue
+		wantJobPodsFinishedMetricValues []metricLabelMapWithValue
+	}{
+		"non-indexed job": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Completions:    pointer.Int32(2),
+					Parallelism:    pointer.Int32(2),
+					CompletionMode: &nonIndexedCompletion,
+				},
+			},
+			completionMode: batchv1.NonIndexedCompletion,
+			wantJobFinishedNumMetricValues: []metricLabelMapWithValue{
+				{
+					LabelsMap: map[string]string{"completion_mode": "NonIndexed", "result": "succeeded"},
+					Value:     1,
+				},
+			},
+			wantJobPodsFinishedMetricValues: []metricLabelMapWithValue{
+				{
+					LabelsMap: map[string]string{"completion_mode": "NonIndexed", "result": "succeeded"},
+					Value:     2,
+				},
+			},
+		},
+		"indexed job": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Completions:    pointer.Int32(2),
+					Parallelism:    pointer.Int32(2),
+					CompletionMode: &indexedCompletion,
+				},
+			},
+			completionMode: batchv1.IndexedCompletion,
+			wantJobFinishedNumMetricValues: []metricLabelMapWithValue{
+				{
+					LabelsMap: map[string]string{"completion_mode": "Indexed", "result": "succeeded"},
+					Value:     1,
+				},
+			},
+			wantJobPodsFinishedMetricValues: []metricLabelMapWithValue{
+				{
+					LabelsMap: map[string]string{"completion_mode": "Indexed", "result": "succeeded"},
+					Value:     2,
+				},
+			},
+		},
+	}
+	job_index := 0 // job index to avoid collisions between job names created by different test cases
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+
+			// record the metrics after the job is created
+			jobFinishedNumBefore, err := getCounterMetricValueForLabels(metrics.JobFinishedNum, tc.wantJobFinishedNumMetricValues)
+			if err != nil {
+				t.Fatalf("Failed to collect the JobFinishedNum metric before creating the job: %q", err)
+			}
+			jobPodsFinishedBefore, err := getCounterMetricValueForLabels(metrics.JobPodsFinished, tc.wantJobFinishedNumMetricValues)
+			if err != nil {
+				t.Fatalf("Failed to collect the JobPodsFinished metric before creating the job: %q", err)
+			}
+
+			// create a single job and wait for its completion
+			job := tc.job.DeepCopy()
+			job.Name = fmt.Sprintf("job-%v", job_index)
+			job_index++
+			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, job)
+			if err != nil {
+				t.Fatalf("Failed to create Job: %v", err)
+			}
+			validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+				Active: int(*jobObj.Spec.Parallelism),
+				Ready:  pointer.Int32(0),
+			}, wFinalizers)
+			if err, _ := setJobPodsPhase(ctx, clientSet, jobObj, v1.PodSucceeded, int(*jobObj.Spec.Parallelism)); err != nil {
+				t.Fatalf("Failed setting phase %s on Job Pod: %v", v1.PodSucceeded, err)
+			}
+			validateJobSucceeded(ctx, t, clientSet, jobObj)
+
+			// record the metrics after the job is finished
```

Comment:
You probably need to use `wait.Poll`, because it's not guaranteed that the metrics are updated right after the jobObj is up-to-date (the job sync could still be running)

Reply:
Indeed. I've never hit the issue until I injected a time.Sleep before recoring the metrics in job_controller. The test started to fail, so it could fail occasionally. Fixed.